File: juno/indicators/rsi.py
from decimal import Decimal

from .smma import Smma


# Relative Strength Index
class Rsi:
    value: Decimal = Decimal('0.0')
    _mean_down: Smma
    _mean_up: Smma
    _last_price: Decimal = Decimal('0.0')
    _t: int = 0
    _t1: int

    def __init__(self, period: int) -> None:
        self._mean_down = Smma(period)
        self._mean_up = Smma(period)
        self._t1 = period + 1

    @property
    def maturity(self) -> int:
        return self._t1

    @property
    def mature(self) -> bool:
        return self._t >= self._t1

    def update(self, price: Decimal) -> Decimal:
        self._t = min(self._t + 1, self._t1)

        if self._t > 1:
            up = Decimal('0.0')
            down = Decimal('0.0')
            if price > self._last_price:
                up = price - self._last_price
            elif price < self._last_price:
                down = self._last_price - price

            self._mean_up.update(up)
            self._mean_down.update(down)

            if self._t == self._t1:
                if self._mean_down.value == 0 and self._mean_up.value != 0:
                    self.value = Decimal('100.0')
                elif self._mean_down.value == 0:
                    self.value = Decimal('0.0')
                else:
                    rs = self._mean_up.value / self._mean_down.value
                    self.value = 100 - (100 / (1 + rs))

        self._last_price = price
        return self.value

File: juno_rs/src/indicators/rsi.rs
use super::{smma::Smma, MA};
use std::cmp::min;

pub struct Rsi {
    pub value: f64,

    mean_down: Smma,
    mean_up: Smma,

    last_price: f64,

    t: u32,
    t1: u32,
}

impl Rsi {
    pub fn new(period: u32) -> Self {
        Self {
            value: 0.0,
            mean_down: Smma::new(period),
            mean_up: Smma::new(period),
            last_price: 0.0,
            t: 0,
            t1: period,
        }
    }

    pub fn maturity(&self) -> u32 {
        self.t1
    }

    pub fn mature(&self) -> bool {
        self.t >= self.t1
    }

    pub fn update(&mut self, price: f64) {
        if self.t > 0 {
            let (up, down) = if price > self.last_price {
                (price - self.last_price, 0.0)
            } else if price < self.last_price {
                (0.0, self.last_price - price)
            } else {
                (0.0, 0.0)
            };

            self.mean_up.update(up);
            self.mean_down.update(down);

            if self.t == self.t1 {
                if self.mean_down.value == 0.0 && self.mean_up.value != 0.0 {
                    self.value = 100.0;
                } else if self.mean_down.value == 0.0 {
                    self.value = 0.0;
                } else {
                    let rs = self.mean_up.value / self.mean_down.value;
                    self.value = 100.0 - (100.0 / (1.0 + rs));
                }
            }
        }

        self.last_price = price;
        self.t = min(self.t + 1, self.t1);
    }
}
